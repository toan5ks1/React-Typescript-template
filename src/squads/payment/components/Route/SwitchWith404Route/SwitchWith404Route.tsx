import { ReactNode } from "react";

import { Redirect, Route, Switch, SwitchProps, useLocation } from "react-router";

interface SwitchWith404RouteProps extends SwitchProps {
    catchAllMode?: "404" | "component";
    component?: ReactNode;
    basename?: string;
}
const SwitchWith404Route = (props: SwitchWith404RouteProps) => {
    const { children, catchAllMode = "404", component } = props;
    const { pathname } = useLocation();
    const shouldRedirect = pathname !== "/";

    return (
        <Switch>
            {children}
            <Route
                render={() => {
                    if (catchAllMode === "404") {
                        if (shouldRedirect) {
                            return <Redirect to="/page-not-found" />;
                        }
                    }

                    if (catchAllMode === "component") return component;
                }}
            />
        </Switch>
    );
};

export default SwitchWith404Route;
