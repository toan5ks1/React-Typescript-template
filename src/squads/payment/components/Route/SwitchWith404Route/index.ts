export { default } from "./SwitchWith404Route";
