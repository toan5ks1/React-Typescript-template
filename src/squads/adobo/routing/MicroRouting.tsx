import { useRouteMatch, Route } from "react-router-dom";

import SwitchWith404Route from "src/components/Route/SwitchWith404Route";

import useInitModules from "./useInitModules";

const MicroRouting = () => {
    const { routes } = useInitModules();
    let { path: rootPath } = useRouteMatch();

    return (
        <SwitchWith404Route>
            {routes.map((route) => {
                const { path } = route;

                return <Route key={path} {...route} path={`${rootPath}${path}`} />;
            })}
        </SwitchWith404Route>
    );
};

export default MicroRouting;
