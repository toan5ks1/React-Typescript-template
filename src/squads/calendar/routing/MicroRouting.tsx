import { useRouteMatch, Route } from "react-router-dom";

import SwitchWith404Route from "src/components/Route/SwitchWith404Route";
import ErrorBoundaryLayout from "src/squads/calendar/components/ErrorBoundary/ErrorBoundaryLayout";
import TranslationProvider from "src/squads/calendar/providers/TranslationProvider";

import useInitModules from "./useInitModules";

const MicroRouting = () => {
    const { routes } = useInitModules();
    let { path: rootPath } = useRouteMatch();

    return (
        <TranslationProvider>
            <ErrorBoundaryLayout>
                <SwitchWith404Route>
                    {routes.map((route) => {
                        const { path } = route;
                        return <Route key={path} {...route} path={`${rootPath}${path}`} />;
                    })}
                </SwitchWith404Route>
            </ErrorBoundaryLayout>
        </TranslationProvider>
    );
};

export default MicroRouting;
