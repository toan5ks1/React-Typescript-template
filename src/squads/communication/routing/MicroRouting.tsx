import { useRouteMatch, Route } from "react-router-dom";

import Loading from "src/components/Loading";
import SwitchWith404Route from "src/components/Route/SwitchWith404Route";
import ErrorBoundaryLayout from "src/squads/communication/components/ErrorBoundary/ErrorBoundaryLayout";
import CommunicationProvider from "src/squads/communication/providers/CommunicationProvider";

import { rawRoutes } from "./routings";

import useCommunicationAppInit from "src/squads/communication/hooks/useCommunicationAppInit";

const MicroRouting = () => {
    let { path: rootPath } = useRouteMatch();

    const { ready, routes } = useCommunicationAppInit(rawRoutes);

    if (!ready) return <Loading fullscreen />;

    return (
        <CommunicationProvider>
            <ErrorBoundaryLayout>
                <SwitchWith404Route>
                    {routes.map((route) => {
                        const { path } = route;
                        return <Route key={path} {...route} path={`${rootPath}${path}`} />;
                    })}
                </SwitchWith404Route>
            </ErrorBoundaryLayout>
        </CommunicationProvider>
    );
};

export default MicroRouting;
