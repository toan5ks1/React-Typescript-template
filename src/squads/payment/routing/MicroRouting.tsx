import { useRouteMatch, Route } from "react-router-dom";

import Loading from "src/components/Loading";
import SwitchWith404Route from "src/components/Route/SwitchWith404Route";
import ErrorBoundaryLayout from "src/squads/payment/components/ErrorBoundary/ErrorBoundaryLayout";

import useInitModules from "./useInitModules";

import TranslationProvider from "src/squads/payment/contexts/TranslationProvider";
import usePaymentUpdatePermissionMicroApplication from "src/squads/payment/hooks/usePaymentUpdatePermissionMicroApplication";

const MicroRouting = () => {
    const { routes } = useInitModules();
    let { path: rootPath } = useRouteMatch();

    return (
        <TranslationProvider>
            <ErrorBoundaryLayout>
                <SwitchWith404Route>
                    {routes.map((route) => {
                        const { path } = route;
                        return <Route key={path} {...route} path={`${rootPath}${path}`} />;
                    })}
                </SwitchWith404Route>
            </ErrorBoundaryLayout>
        </TranslationProvider>
    );
};

const MicroRoutingWithPermission = () => {
    const role = usePaymentUpdatePermissionMicroApplication();
    if (!role) return <Loading fullscreen />;

    return <MicroRouting />;
};
export default MicroRoutingWithPermission;
