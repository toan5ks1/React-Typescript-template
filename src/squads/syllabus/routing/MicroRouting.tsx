import { useRouteMatch, Route } from "react-router-dom";

import ErrorBoundaryLayout from "../components/ErrorBoundary/ErrorBoundaryLayout";
import Loading from "src/components/Loading";
import SwitchWith404Route from "src/components/Route/SwitchWith404Route";
import TranslationProvider from "src/squads/syllabus/providers/TranslationProvider";

import useInitModules from "./useInitModules";

import useSyllabusUpdatePermissionMicroApplication from "src/squads/syllabus/hooks/useSyllabusUpdatePermissionMicroApplication";

const MicroRouting = () => {
    const { routes } = useInitModules();
    let { path: rootPath } = useRouteMatch();

    return (
        <TranslationProvider>
            <ErrorBoundaryLayout>
                <SwitchWith404Route>
                    {routes.map((route, index) => {
                        const { path } = route;
                        return (
                            <Route
                                key={`${path}-${index}`}
                                {...route}
                                path={`${rootPath}${path}`}
                            />
                        );
                    })}
                </SwitchWith404Route>
            </ErrorBoundaryLayout>
        </TranslationProvider>
    );
};

const MicroRoutingWithPermission = () => {
    const role = useSyllabusUpdatePermissionMicroApplication();
    if (!role) return <Loading fullscreen />;

    return <MicroRouting />;
};
export default MicroRoutingWithPermission;
