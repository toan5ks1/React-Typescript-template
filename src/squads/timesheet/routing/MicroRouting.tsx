import { useRouteMatch, Route } from "react-router-dom";

import Loading from "src/components/Loading";
import SwitchWith404Route from "src/components/Route/SwitchWith404Route";
import ErrorBoundaryLayout from "src/squads/timesheet/components/ErrorBoundary/ErrorBoundaryLayout";
import TranslationProvider from "src/squads/timesheet/providers/TranslationProvider";

import useInitModules from "./useInitModules";

import useTimesheetUpdatePermissionMicroApplication from "src/squads/timesheet/hooks/useTimesheetUpdatePermissionMicroApplication";

const MicroRouting = () => {
    const { routes } = useInitModules();
    let { path: rootPath } = useRouteMatch();

    return (
        <TranslationProvider>
            <ErrorBoundaryLayout>
                <SwitchWith404Route>
                    {routes.map((route) => {
                        const { path } = route;

                        return <Route key={path} {...route} path={`${rootPath}${path}`} />;
                    })}
                </SwitchWith404Route>
            </ErrorBoundaryLayout>
        </TranslationProvider>
    );
};

const MicroRoutingWithPermission = () => {
    const role = useTimesheetUpdatePermissionMicroApplication();
    if (!role) return <Loading fullscreen />;

    return <MicroRouting />;
};
export default MicroRoutingWithPermission;
