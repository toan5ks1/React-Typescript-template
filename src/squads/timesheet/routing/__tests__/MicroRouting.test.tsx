import { Suspense } from "react";

import { MemoryRouter, Route } from "react-router";

import SwitchWith404Route from "src/components/Route/SwitchWith404Route";

import MicroRouting from "../MicroRouting";

import { render } from "@testing-library/react";

jest.mock("src/squads/timesheet/internals/permission", () => {
    const originalModule = jest.requireActual("src/squads/timesheet/internals/permission");
    return {
        ...originalModule,
        can: () => true,
    };
});
jest.mock("src/app/useFeatureController");

jest.mock("react-router", () => {
    const actual = jest.requireActual("react-router");
    return {
        __esModule: true,
        ...actual,
        useRouteMatch: () => ({
            path: "/timesheet",
        }),
    };
});
jest.mock("src/squads/timesheet/routing/useInitModules", () => {
    return {
        __esModule: true,
        default: () => ({
            routes: [
                {
                    path: "/timesheet",
                    component: () => <div>TimesheetRouter</div>,
                },
            ],
        }),
    };
});

jest.mock("src/squads/timesheet/hooks/useTimesheetUpdatePermissionMicroApplication", () => {
    return {
        __esModule: true,
        default: () => "ROLE",
    };
});

describe("<MicroRouting />", () => {
    it("should render /timesheet path", () => {
        const wrapper = render(
            <MemoryRouter initialEntries={["/timesheet/timesheet"]}>
                <Suspense fallback={"..."}>
                    <SwitchWith404Route>
                        <Route path="/timesheet" component={MicroRouting} />
                    </SwitchWith404Route>
                </Suspense>
            </MemoryRouter>
        );
        expect(wrapper.container).toMatchSnapshot();
    });
});
