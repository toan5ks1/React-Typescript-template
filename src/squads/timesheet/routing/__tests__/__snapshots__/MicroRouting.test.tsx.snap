// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MicroRouting /> should render /timesheet path 1`] = `
<div>
  <div>
    TimesheetRouter
  </div>
</div>
`;
